examples/datalog/stratifier.flix

```diff
@@ -0,0 +1,88 @@
+namespace Stratifier {
+    type alias Program = List[Constraint]
+    type alias Constraint = (HeadAtom, List[BodyAtom])
+    type alias Atom = PredicateSymbol
+    type alias HeadAtom = Atom
+    enum BodyAtom {
+        case Positive(Atom)
+        case Negative(Atom)
+    }
+    type alias PredicateSymbol = String
+
+    // A :- B leads to (A, true, B)
+    type alias PrecedenceEdge = (PredicateSymbol, Bool, PredicateSymbol)
+    type alias PrecedenceGraph = List[PrecedenceEdge]
+
+    pub def convertProgram(p: Program): PrecedenceGraph =
+        List.flatMap(convertConstraint, p)
+
+    pub def convertConstraint(c: Constraint): PrecedenceGraph =
+        use List.map;
+        let (a0, b) = c;
+        let unfoldBodyAtom = ba -> match ba {case Positive(a) => (true, a), case Negative(a) => (false, a)};
+        b |> map(unfoldBodyAtom) |> map(pna -> match pna {case (pn, a) => (a0, pn, a)})
+
+    pub def ullman(pg: PrecedenceGraph, predicates: Int): Option[Map[PredicateSymbol, Int]] & Impure =
+        use Int32.max;
+        let facts = project pg into Precedence;
+        let rules = #{
+            Predicate(p; 1) :- Precedence(p, _, _).
+            Predicate(p; 1) :- Precedence(_, _, p).
+            Predicate(ph; max(pbs, phs)) :- Precedence(ph, true, pb), Predicate(pb; pbs), Predicate(ph; phs).
+            Predicate(ph; max(pbs + 1, phs)) :-
+                Precedence(ph, false, pb),
+                Predicate(pb; pbs),
+                Predicate(ph; phs),
+                if max(pbs + 1, phs) <= predicates.
+            NonStratified() :-
```

Comment:
Yes, this can be done, and it uses Datalog "more". But I find it cumbersome.

Why not let m = `select (p, s) from Predicate` |> toMap...

let stratified = Map.exists((_, stratum) -> stratum > predicates)

if (stratified) ...